packages/web3torrent/src/hooks/use-budget.ts

```diff
@@ -2,21 +2,42 @@ import {web3TorrentClient} from '../clients/web3torrent-client';
 import {INITIAL_BUDGET_AMOUNT} from '../constants';
 import {useState, useEffect} from 'react';
 import {DomainBudget} from '@statechannels/client-api-schema';
+import {from, noop, Subscription} from 'rxjs';
+import {safeUnsubscribe, safeUnsubscribeFromFunction} from '../utils/react-utls';
+import {logger} from '../logger';
+
+const log = logger.child({module: 'use-budget'});
 
 export function useBudget({ready}: {ready: boolean}) {
   const {paymentChannelClient} = web3TorrentClient;
 
   const [budget, setBudget] = useState<DomainBudget>(undefined);
   const [loading, setLoading] = useState(true);
   useEffect(() => {
-    const getAndSetBudget = async () => {
-      const budget = await paymentChannelClient.getBudget();
-
+    const setBudgetUI = budget => {
       setBudget(budget);
       setLoading(false);
     };
-    if (ready) getAndSetBudget();
-    return paymentChannelClient.channelClient.onBudgetUpdated(getAndSetBudget);
+    const getBudget = () => from(paymentChannelClient.getBudget());
+
+    let getBudgetSubscription: Subscription;
+    if (ready) {
+      getBudgetSubscription = getBudget().subscribe(setBudgetUI);
+    }
+    const onBudgetUpdatedSubscription = paymentChannelClient.channelClient.onBudgetUpdated(() =>
+      getBudget().subscribe(setBudgetUI)
+    );
+
+    // Mostly works.
+    // Cleanup is not properly done for the following case:
+    // 1. onBudgetUpdated event arrives.
+    // 2. getBudget().subscribe(setBudget) is invoked.
+    // The operation from step 2 is not properly cleaned up.
+    // To fix this issue, onBudgetUpdated would need to return an observable instead of a subscription.
```

Comment:
Would something like this work?
```ts
useEffect(() => {
  let cancelled = false;
  const getAndSetBudget = async () => {
    const budget = await paymentChannelClient.getBudget();

    !cancelled && setBudget(budget);
    !cancelled && setLoading(false);
  };
  if (ready) getAndSetBudget();
  const unsubscribe = paymentChannelClient.channelClient.onBudgetUpdated(getAndSetBudget);

  return () => { cancelled = true; unsubscribe() };
})
```

Reply:
I like that idea. Implemented in 0c0f8af349a0b735adef3e08f27cd2e8bf03d086